formats/protobuf/src/main/java/io/cloudevents/protobuf/ProtoSerializer.java

```diff
@@ -0,0 +1,261 @@
+import io.cloudevents.rw.CloudEventRWException;
+import io.cloudevents.v1.proto.CloudEvent;
+import io.cloudevents.v1.proto.CloudEvent.*;
+
+import java.net.URI;
+import java.time.Instant;
+import java.time.OffsetDateTime;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import static io.cloudevents.protobuf.ProtobufFormat.PROTO_DATA_CONTENT_TYPE;
+import static io.cloudevents.v1.proto.CloudEvent.*;
+
+/**
+ * Provides functionality for turning a {@link io.cloudevents.CloudEvent} to the protobuf representation {@link CloudEvent}.
+ */
+class ProtoSerializer {
+
+    /**
+     * Convert the Java SDK CloudEvent into a protobuf representation.
+     *
+     * <p> If the data payload is set with "datacontenttype" set to "application/protobuf", the data provided
+     * must be wrapped in {@link com.google.protobuf.Any}. Storing protobuf data in an Any is required by the
+     * CloudEvent protobuf format spec.
+     *
+     * @param ce The {@link io.cloudevents.CloudEvent} to convert.
+     * @return A {@link CloudEvent} protobuf object that can be serialized to the protobuf binary format.
+     * @throws InvalidProtocolBufferException If the content type of the data is "application/protobuf" but
+     *  the data is not wrapped in {@link com.google.protobuf.Any}.
+     */
+    public static CloudEvent toProto(io.cloudevents.CloudEvent ce) throws InvalidProtocolBufferException {
+        CloudEvent.Builder builder = CloudEvent.newBuilder();
+
+        builder.setSpecVersion(ce.getSpecVersion().toString());
+
+        // Copy the attributes
+        ProtoContextWriter protoContextWriter = new ProtoContextWriter(builder, ce.getSpecVersion());
+        final CloudEventContextReader cloudEventContextReader = CloudEventUtils.toContextReader(ce);
+        cloudEventContextReader.readContext(protoContextWriter);
+
+        // Copy the data
+        final String contentType = ce.getDataContentType();
+        final CloudEventData data = ce.getData();
+        if (data != null) {
+
+            // If it's a proto message we can handle that directly.
+
+            if (data instanceof ProtoCloudEventData) {
+
+                final ProtoCloudEventData protoData = (ProtoCloudEventData) data;
+
+                if (protoData.getMessage() != null) {
+                    builder.setProtoData(Any.pack(protoData.getMessage()));
+                }
+
+            } else {
+                if (Objects.equals(contentType, PROTO_DATA_CONTENT_TYPE)) {
+                    // This will throw if the data provided is not an Any. The protobuf CloudEvent spec requires proto data to be stored as
+                    // an Any. I would be amenable to allowing people to also pass raw unwrapped protobufs, but it complicates the logic here.
+                    // Perhpas that can be a follow up if there is a need.
+                    Any dataAsAny = Any.parseFrom(data.toBytes());
```

Comment:
Any is whatever message encoded as protobuf I guess? If that's the case, I think this logic is fine, but users should be able to wrap in a `CloudEventData` an `Any`. I guess the way to do that is to have a static method in `ProtoCloudEventData` to create a `ProtoCloudEventData`, so it will fall in the other if branch?

Reply:
Ack..